### PR TITLE
State Refactor Complete

### PR DESCRIPTION
**Pull request by ukstv**
Proper concurrency model in place.

### CODE REVIEW

File: packages/core/src/state-management/execution-queue.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import { Task, TaskQueueLike } from '../pubsub/task-queue';
+import { DiagnosticsLogger } from '@ceramicnetwork/logger';
+import { RunningState } from './running-state';
+
+export interface ExecLike extends TaskQueueLike {
+  addE: (task: (state$: RunningState) => Promise<void>) => void;
+  runE<T>(task: (state$: RunningState) => Promise<T>): Promise<T>;
+}
+
+export class ExecutionQueue {
```
Comment (stbrody) [inline packages/core/src/state-management/execution-queue.ts:12]: Can you add some comments on the relationship and differences between TaskQueue and ExecutionQueue?

File: packages/core/src/state-management/execution-queue.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import { NamedTaskQueue } from './named-task-queue';
+import { DocID } from '@ceramicnetwork/docid';
+import { Task, TaskQueueLike } from '../pubsub/task-queue';
+import { DiagnosticsLogger } from '@ceramicnetwork/logger';
+import { RunningState } from './running-state';
+
+export interface ExecLike extends TaskQueueLike {
+  addE: (task: (state$: RunningState) => Promise<void>) => void;
```
Comment (stbrody) [inline packages/core/src/state-management/execution-queue.ts:8]: If you have an ExecutionQueue, when would you call `.addE` or `runE` vs `.add` or `.run`?  Some comments here would help

File: packages/core/src/store/level-state-store.ts [hunk 3]
```diff
@@ -36,7 +36,7 @@ export class LevelStateStore implements StateStore {
      * @param docStateHolder - Document instance
      */
     async save(docStateHolder: DocStateHolder): Promise<void> {
-        await this.#store.put(docStateHolder.id.baseID.toString(), DoctypeUtils.serializeState(docStateHolder.state))
+        await this.#store.put(docStateHolder.id.toString(), DoctypeUtils.serializeState(docStateHolder.state))
```
Comment (stbrody) [inline packages/core/src/store/level-state-store.ts:39]: why this change?  Don't we always want to store by baseID?  I'm not sure it makes sense to cache state at specific commitIds in the state store.

Maybe we should just throw if the doc is pinned to a specific CommitID when trying to store it into the state store?

File: packages/core/src/state-management/repository.ts [hunk 4]
```diff
@@ -1,70 +1,110 @@
...
+  anchorService: AnchorService,
+  conflictResolution: ConflictResolution,
+  stateValidation: StateValidation
+}
 
 export class Repository {
-  readonly loadingQ: NamedTaskQueue = new NamedTaskQueue();
-  readonly #map: AsyncLruMap<Document>;
-  #documentFactory?: DocumentFactory;
+  readonly loadingQ: NamedTaskQueue;
```
Comment (stbrody) [inline packages/core/src/state-management/repository.ts:27]: I think it'd be helpful if all of these members of Repository had brief comments on what they represent

File: packages/core/src/state-management/repository.ts [hunk 5]
```diff
@@ -98,62 +138,44 @@ export class Repository {
...
     } else {
-      if (this.pinStore) {
-        return this.pinStore.stateStore.load(docId);
-      }
+      return this.#deps.pinStore.stateStore.load(docId);
     }
   }
 
   /**
    * Stub for adding the document.
```
Comment (stbrody) [inline packages/core/src/state-management/repository.ts:150]:
```suggestion
   * Adds the document to the in-memory cache
```

File: packages/core/src/state-management/snapshot-state.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { Observable, of } from 'rxjs';
+import { RunningStateLike } from './running-state';
+import { DocState } from '@ceramicnetwork/common';
+import { DocID } from '@ceramicnetwork/docid';
+
+export class SnapshotState extends Observable<DocState> implements RunningStateLike {
```
Comment (stbrody) [inline packages/core/src/state-management/snapshot-state.ts:6]: class comment please

File: packages/core/src/state-management/state-manager.ts [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+   * the most recent Tip to it.
+   * @param state$
+   * @param opts
+   */
+  syncGenesis(state$: RunningState, opts: DocOpts): Promise<void> {
+    return this.applyOpts(state$, opts);
+  }
+
+  /**
+   * Takes the most recent known-about version of a document and a specific commit and returns a new
```
Comment (stbrody) [inline packages/core/src/state-management/state-manager.ts:38]: Comment is out of date now

**Review by stbrody — CHANGES_REQUESTED** [review]
This looks really great!
Most of my suggestions are around adding more comments and clarifying the names of various components and their relationships to other components.  But the core of the change looks great, and I think overall the code looks much cleaner now.  Thanks for all your hard work here!

**Comment by stbrody** [comment]
One other question: As best as I can tell this hasn't changed the behavior around how and when new states become visible to doctype instances.  I assume there will be another PR coming that implements the changes that we discussed in https://www.notion.so/threebox/User-visible-behavior-on-document-updates-382c1c8e24b040f2a54b78b06a2466f9?

File: packages/core/src/ceramic.ts [hunk 8]
```diff
@@ -629,22 +645,25 @@ class Ceramic implements CeramicApi {
...
-      const commitId = doc.findCommitAt(opts.atTime)
-      return doc.rewind(commitId)
+      return this.repository.stateManager.atTime(state$, opts.atTime)
     } else {
-      return doc
+      return state$
     }
   }
 
+  async __loadDoc(docId: DocID, opts: DocOpts = {}): Promise<RunningState> {
```
[the remark this answers is not in the document]
Comment (oed) [reply packages/core/src/ceramic.ts:663]: Also don't get why both are needed 🤔

File: packages/core/src/conflict-resolution.ts [hunk 9]
```diff
@@ -205,18 +207,33 @@ export async function fetchLog(
   return fetchLog(dispatcher, prevCid, stateLog, log);
 }
 
+export function commitAtTime(state$: RunningStateLike, timestamp: number): CommitID {
```
Comment (oed) [inline packages/core/src/conflict-resolution.ts:210]: 👍

File: packages/core/src/store/level-state-store.ts (hunk 3 above)
Comment (oed) [reply packages/core/src/store/level-state-store.ts:39]: Storing states for specific CommitIDs in the state store might make sense when you are loading a commit of a document multiple times. This happens in 3ID and for schemas for example.

**Review by oed — COMMENTED** [review]
Looks good :)

File: packages/core/src/store/level-state-store.ts (hunk 3 above)
Comment (stbrody) [reply packages/core/src/store/level-state-store.ts:39]: But you could just store the current state of the doc in the state store, load it, then reset to the old version

Comment (oed) [reply packages/core/src/store/level-state-store.ts:39]: That's going to be quite slow if you are loading it multiple times.

Comment (oed) [reply packages/core/src/store/level-state-store.ts:39]: Ideally that's the case yes. I would guess that's not how it's implemented right now however.

File: packages/core/src/__tests__/document.test.ts [hunk 10]
```diff
@@ -1,29 +1,21 @@
-import { DocID } from '@ceramicnetwork/docid';
-import { AnchorStatus, CommitType, DocState, DoctypeUtils, IpfsApi, SignatureStatus } from '@ceramicnetwork/common';
+import {
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/__tests__/document.test.ts:1]: True.

File: packages/core/src/ceramic.ts [hunk 11]
```diff
@@ -209,12 +208,19 @@ class Ceramic implements CeramicApi {
 
     this._doctypeHandlers = new HandlersMap(this._logger)
 
+    // This initialization block below has to be redone.
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/ceramic.ts:211]: Changed.

File: packages/core/src/ceramic.ts [hunk 12]
```diff
@@ -209,12 +208,19 @@ class Ceramic implements CeramicApi {
...
     this.stateValidation = this._validateDocs ? new RealStateValidation(this.loadDocument.bind(this)) : new FauxStateValidation()
-    const documentFactory = new DocumentFactory(this.dispatcher, pinStore, this.context, this._doctypeHandlers, this.stateValidation)
-    const networkLoad = new NetworkLoad(this.dispatcher, this._doctypeHandlers, this.context, this._logger, documentFactory)
-    this.repository.setPinStore(pinStore)
-    this.repository.setDocumentFactory(documentFactory);
-    this.repository.setNetworkLoad(networkLoad);
+    const conflictResolution = new ConflictResolution(this.context.anchorService, this.stateValidation, this.dispatcher, this.context, this._doctypeHandlers)
+    const pinStore = modules.pinStoreFactory.createPinStore()
+    this.repository.setDeps({
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/ceramic.ts:216]: My position is Context is not for DI. For example, AnchorService should not be a part of Context IMO.

[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/ceramic.ts:216]: Discussion about Context: worth it.

File: packages/core/src/ceramic.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/ceramic.ts]: Forgotten remnant of dev time. Only one is required.

[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/ceramic.ts]: Changed.

File: packages/core/src/ceramic.ts [hunk 13]
```diff
@@ -629,22 +645,25 @@ class Ceramic implements CeramicApi {
    * @param docId - Document ID
    * @param opts - Initialization options
    */
-  async _loadDoc(docId: DocID | CommitID | string, opts: DocOpts = {}): Promise<Document> {
+  async _loadDoc(docId: DocID | CommitID | string, opts: DocOpts = {}): Promise<RunningStateLike> {
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/ceramic.ts]: Renamed `__loadDoc` as `_loadDoc` with old `_loadDoc` gone. Hope now it is clearer.

File: packages/core/src/ceramic.ts [hunk 14]
```diff
@@ -656,16 +675,17 @@ class Ceramic implements CeramicApi {
...
-      const toRecover = document.state?.anchorStatus === AnchorStatus.PENDING || document.state?.anchorStatus === AnchorStatus.PROCESSING
-      if (toRecover) {
-        document.anchor()
-      }
+  restoreDocuments() {
+    this.repository.listPinned().then(async list => {
+      let n = 0
+      await Promise.all(list.map(async docId => {
+        await this.__loadDoc(DocID.fromString(docId))
+        n++;
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/ceramic.ts:669]: ¯\_(ツ)_/¯

File: packages/core/src/conflict-resolution.ts [hunk 15]
```diff
@@ -205,18 +207,33 @@ export async function fetchLog(
...
-    private readonly handler: ContextfulHandler<Doctype>,
+    private readonly context: Context,
+    private readonly handlers: HandlersMap
   ) {}
 
   /**
    * Applies the log to the document and updates the state.
    */
-  async applyLogToState(log: Array<CID>, state?: DocState, breakOnAnchor?: boolean): Promise<DocState> {
+  private async applyLogToState(handler: ContextfulHandler, log: Array<CID>, state?: DocState, breakOnAnchor?: boolean): Promise<DocState> {
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/conflict-resolution.ts:236]: Uhums.

[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/conflict-resolution.ts]: Too lazy to add ContextfulHandlersMap instead :)

File: packages/core/src/state-management/execution-queue.ts (hunk 2 above)
Comment (ukstv) [reply packages/core/src/state-management/execution-queue.ts:8]: Replaced `addE`/`runE` with just `add`/`run`. Sans-E functions are not used anyway.

File: packages/core/src/store/level-state-store.ts (hunk 3 above)
Comment (ukstv) [reply packages/core/src/store/level-state-store.ts:39]: Surprised about this. `DocStateHolder#id` is defined as `DocID` type. `DocID#baseID` is just redundant here.

File: packages/core/src/state-management/execution-queue.ts (hunk 1 above)
Comment (ukstv) [reply packages/core/src/state-management/execution-queue.ts:29]: Added comments to the ExecutionQueue.

File: packages/core/src/state-management/repository.ts (hunk 4 above)
Comment (ukstv) [reply packages/core/src/state-management/repository.ts:27]: Ok.

File: packages/core/src/state-management/snapshot-state.ts (hunk 6 above)
Comment (ukstv) [reply packages/core/src/state-management/snapshot-state.ts:6]: Okay.

File: packages/core/src/state-management/state-manager.ts (hunk 7 above)
Comment (ukstv) [reply packages/core/src/state-management/state-manager.ts:38]: Changed.

**Comment by ukstv** [comment]
"One other question: As best as I can tell this hasn't changed the behavior around how and when new states become visible to doctype instances"

Even more so, this does not change how doctype relates to state. I extracted the functionality to https://github.com/ceramicnetwork/js-ceramic/pull/1141/files#diff-d32156236fb949a52abdd6b3238467d9ee30f43a96eb746de316dae81304c995R5 to track this easier. It is the next step to change Doctype.

File: packages/core/src/store/level-state-store.ts (hunk 3 above)
Comment (stbrody) [reply packages/core/src/store/level-state-store.ts:39]: > Ideally that's the case yes. I would guess that's not how it's implemented right now however.

Shouldn't the plan then be to move in the direction towards optimizing the implementation rather than attempting to optimize by storing more data? @oed

Comment (stbrody) [reply packages/core/src/store/level-state-store.ts:39]: > Surprised about this. `DocStateHolder#id` is defined as `DocID` type. `DocID#baseID` is just redundant here.

Ah right, sorry, my mistake.

How do we ensure that we aren't putting the state from a "readonly" document at was loaded at a specific CommitID into the state store as if it was the most recent version of that document?  Can we add an assert somewhere into this process that would catch that type of programming or user error?  It would be a pretty nasty bug if we ever accidentally overrode the state store with less up-to-date info for a given DocID than what was already there

File: packages/core/src/ceramic.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (stbrody) [reply packages/core/src/ceramic.ts:216]: Yep.  Your comment on the other PR about how using Context reduces visibility into which parts of the Context are actually needed by any given piece of code is a good point.  But then if Context isn't for dependency injection it raises the question: what *is* Context for?  Should we be moving towards eliminating Context entirely?  If so, should we be replacing it with some other type of DI system?  Or should we just get used to passing around a lot of references everywhere to what are basically process-wide singleton objects?

In any case, this seems like the topic for a separate more dedicated conversation about our approach to dependency injection in this codebase.  Which I think can safely wait until after mainnet is out the door

File: packages/core/src/state-management/execution-queue.ts [hunk 16]
```diff
@@ -1,14 +1,31 @@
 import { NamedTaskQueue } from './named-task-queue';
```
Comment (stbrody) [inline packages/core/src/state-management/execution-queue.ts:1]: Nice cleanup of this file over the last version, I find it much easier to make sense of now!

**Review by stbrody — COMMENTED** [review]
> It is the next step to change Doctype.

Perfect!

**Review by stbrody — APPROVED** [review]
Looks great!  Thanks for all the nice cleanups!

File: packages/core/src/state-management/execution-queue.ts (hunk 16 above)
Comment (ukstv) [reply packages/core/src/state-management/execution-queue.ts:1]: ^_^

File: packages/core/src/store/level-state-store.ts (hunk 3 above)
Comment (ukstv) [reply packages/core/src/store/level-state-store.ts:39]: @stbrody I believe, this check could be made on type level. Right now all of RunningState, RunningStateLike, SnapshotState along with Doctype implement DocStateHolder interface. Maybe at the next iteration, we think on proper typing to avoid this kind of accident.
